runtime/tests/checker/account_test.go

```diff
@@ -1644,3 +1644,242 @@ func TestCheckAccountPaths(t *testing.T) {
 		require.NoError(t, err)
 	})
 }
+
+func TestCheckPublicAccountIteration(t *testing.T) {
+
+	t.Parallel()
+
+	t.Run("basic", func(t *testing.T) {
+		t.Parallel()
+		_, err := ParseAndCheckAccount(t,
+			`
+			fun test() {
+				publicAccount.forEachPublic(fun (path: PublicPath, type:Type): Bool {
+					return true
+				})
+			}
+			`,
+		)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("labels irrelevant", func(t *testing.T) {
+		t.Parallel()
+		_, err := ParseAndCheckAccount(t,
+			`
+			fun test() {
+				publicAccount.forEachPublic(fun (foo: PublicPath, bar:Type): Bool {
+					return true
+				})
+			}
+			`,
+		)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("incompatible return", func(t *testing.T) {
+		t.Parallel()
+		_, err := ParseAndCheckAccount(t,
+			`
+			fun test() {
+				publicAccount.forEachPublic(fun (path: PublicPath, type:Type): Bool {
+					return 3
+				})
+			}
+			`,
+		)
+
+		require.Error(t, err)
+		errors := ExpectCheckerErrors(t, err, 1)
+		require.IsType(t, &sema.TypeMismatchError{}, errors[0])
+	})
+
+	t.Run("incompatible return annot", func(t *testing.T) {
+		t.Parallel()
+		_, err := ParseAndCheckAccount(t,
+			`
+			fun test() {
+				publicAccount.forEachPublic(fun (path: PublicPath, type:Type): Void {})
+			}
+			`,
+		)
+
+		require.Error(t, err)
+		errors := ExpectCheckerErrors(t, err, 1)
+		require.IsType(t, &sema.TypeMismatchError{}, errors[0])
+	})
+
+	t.Run("incompatible arg 1", func(t *testing.T) {
+		t.Parallel()
+		_, err := ParseAndCheckAccount(t,
+			`
+			fun test() {
+				publicAccount.forEachPublic(fun (path: StoragePath, type:Type): Void {})
+			}
+			`,
+		)
+
+		require.Error(t, err)
+		errors := ExpectCheckerErrors(t, err, 1)
+		require.IsType(t, &sema.TypeMismatchError{}, errors[0])
+	})
+
+	t.Run("incompatible arg 2", func(t *testing.T) {
+		t.Parallel()
+		_, err := ParseAndCheckAccount(t,
+			`
+			fun test() {
+				publicAccount.forEachPublic(fun (path: PublicPath, type:Int): Void {})
+			}
+			`,
+		)
+
+		require.Error(t, err)
+		errors := ExpectCheckerErrors(t, err, 1)
+		require.IsType(t, &sema.TypeMismatchError{}, errors[0])
+	})
+
+	t.Run("supertype", func(t *testing.T) {
+		t.Parallel()
+		_, err := ParseAndCheckAccount(t,
+			`
+			fun test() {
+				publicAccount.forEachPublic(fun (path: CapabilityPath, type:Type): Void {})
+			}
+			`,
+		)
+
+		require.Error(t, err)
+		errors := ExpectCheckerErrors(t, err, 1)
+		require.IsType(t, &sema.TypeMismatchError{}, errors[0])
+	})
+}
+
+func TestCheckAuthAccountIteration(t *testing.T) {
+
+	t.Parallel()
+
+	t.Run("basic suite", func(t *testing.T) {
+		t.Parallel()
+
+		nameTypePairs := []struct {
+			name        string
+			correctType string
+		}{
+			{name: "forEachPublic", correctType: "PublicPath"},
+			{name: "forEachPrivate", correctType: "PrivatePath"},
+			{name: "forEachStored", correctType: "StoragePath"},
+		}
+
+		for _, pair := range nameTypePairs {
+			t.Run("basic", func(t *testing.T) {
+				t.Parallel()
```

Comment:
Careful: Using `t.Parallel` will pause the closure, then restart it. By the time it's restarted, the `pair` variable of the outer for-loop has been rebound. This basically results in totally random behaviour, in the worst case only  the last element of `nameTypePairs` is tested 3 times.

You can fix this by factoring the body of the loop into a helper function and then calling it, e.g. (pseudo-code)

```go
test := func(pair /* ... */) { 
	t.Run(fmt.Sprintf("basic: %s", pair.name), func(t *testing.T) {
		t.Parallel()
		// ...
	})

	// ... 
}

for _, pair := range nameTypePairs {
	test(pair)
}
```

This rebinds `pair`.

Also, the current loop runs a sub-test "basic" for each function, maybe use the pair's name in it, otherwise you get generated subtest names like `basic #0`, `basic #1`, etc.